### ISSUE TITLE
Whisper Encoder issues with Executor API

### ISSUE THREAD

**Issue by MahmoudAshraf97** [NVIDIA/TensorRT-LLM#2338]
Hello, `0.15.0.dev2024101500` added a new issue when using the executor API with whisper

```
[TensorRT-LLM][ERROR] IExecutionContext::inferShapes: Error Code 7: Internal Error (WhisperEncoder/__add___L322/elementwise_binary_L2856/ELEMENTWISE_SUM_0: dimensions not compatible for elementwise. Broadcast has incompatible dimensions: 3000 != 1500 && 3000 != 1 && 1500 != 1. Instruction: CHECK_BROADCAST 3000 1500.)
[TensorRT-LLM][ERROR] Encountered an error in forwardAsync function: Invalid input shape (/home/jenkins/agent/workspace/LLM/main/L0_PostMerge/tensorrt_llm/cpp/tensorrt_llm/runtime/tllmRuntime.cpp:313)
```

This error was introduced after the change in positional embeddings in the last update, when commenting the problematic addition here:
https://github.com/NVIDIA/TensorRT-LLM/blob/75057cd036af25e288c004d8ac9e52fd2d6224aa/tensorrt_llm/models/enc_dec/model.py#L1928

the error is solved, but new error comes up which is also reproducible in `0.14.0.dev2024100800`
```
[TensorRT-LLM][ERROR] Encountered an error in forwardAsync function: slice 1152000 exceeds buffer size 576000
```
this error occurs in all whisper models, the slice is always double the buffer size no matter what the model size is

### Who can help?

@byshiue 

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```bash
INFERENCE_PRECISION=float16
WEIGHT_ONLY_PRECISION=int8
MAX_BEAM_WIDTH=2
MAX_BATCH_SIZE=8

trtllm-build  --checkpoint_dir ${checkpoint_dir}/encoder \
              --output_dir ${output_dir}/encoder \
              --kv_cache_type paged \
              --moe_plugin disable \
              --enable_xqa disable \
              --max_batch_size ${MAX_BATCH_SIZE} \
              --gemm_plugin disable \
              --bert_attention_plugin ${INFERENCE_PRECISION} \
              --max_input_len 3000 \
              --max_seq_len 3000 

trtllm-build  --checkpoint_dir ${checkpoint_dir}/decoder \
              --output_dir ${output_dir}/decoder \
              --kv_cache_type paged \
              --moe_plugin disable \
              --enable_xqa enable \
              --max_beam_width ${MAX_BEAM_WIDTH} \
              --max_batch_size ${MAX_BATCH_SIZE} \
              --max_seq_len 114 \
              --max_input_len 14 \
              --max_encoder_input_len 1500 \
              --gemm_plugin ${INFERENCE_PRECISION} \
              --bert_attention_plugin ${INFERENCE_PRECISION} \
              --gpt_attention_plugin ${INFERENCE_PRECISION} 
```
```python
import tensorrt_llm.bindings.executor as trtllm
import torch

# Create the executor.
executor = trtllm.Executor(
    encoder_model_path=engine_dir + "/encoder/",
    decoder_model_path=engine_dir + "/decoder/",
    model_type=trtllm.ModelType.ENCODER_DECODER,
    executor_config=trtllm.ExecutorConfig(1, gpu_weights_percent=0.5),
)

request = trtllm.Request(
    input_token_ids=[50257, 50258, 50358, 50362],
    max_tokens=100,
    encoder_input_features=torch.randn(3000,80).half().cuda(), # mel features
)

# Enqueue the request.
request_id = executor.enqueue_request(request)
```
</details>

### Expected behavior

It should complete the request successfuly

### additional notes

The decoder works perfectly with executor API but loading the encoder alone or the encoder and the decoder together is not working

**Comment by yuekaizhang** [comment]
@MahmoudAshraf97  You should set `encoder_output_length` https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/runtime/model_runner_cpp.py#L625.

**Comment by MahmoudAshraf97** [comment]
That solved it thanks, one more question, how can I return the encoder output along with the generated tokens? encoder output is used in whisper to generate word timestamps

**Comment by yuekaizhang** [comment]
@MahmoudAshraf97  I have not tried that yet. You may check here https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/runtime/model_runner_cpp.py#L573. Would you mind updating here if it could work?

**Comment by MahmoudAshraf97** [comment]
It did return a tensor, but that tensor is always zeros regardless of the input.

**Comment by yuekaizhang** [comment]
@symphonylyh  Do you have some ideas about how to get encoder_out?

**Comment by MahmoudAshraf97** [comment]
BTW, `log_probs` is also zeros, generation and context logits work fine

**Comment by karljang** [comment]
@MahmoudAshraf97 ,
Apologies for the long silence from the team, let me try to fill in the gap~
I'm sharing a possible solution below, generated by my AI tool for you. 
Also, here are the results from the `tiny` whisper model I tested using`tensorrt-llm 1.2.0rc1` : 

> "Input [Text 0]: "<|startoftranscript|><|en|><|transcribe|><|notimestamps|>"
> Output [Text 0 Beam 0]: " Yet these thoughts affected hester prin less with hope than apprehension. Yet these thoughts affected hester prin less with hope than apprehension. Yet these thoughts affected hester prin less with hope than apprehension. Yet these thoughts affected hester prin less with hope,""

**One additional important note**: since the 1.0 release, TensorRT-LLM has transitioned to the PyTorch workflow, which no longer relies on trtllm-build. This also means we won’t be adding new features to the legacy TensorRT workflows 😄  That said, I don’t yet have enough knowledge to enable Whisper using the PyTorch workflow—but we're always open to contribution!



## 🔍 Root Causes

Your build commands had three main issues:

1. **Wrong flag name**: Using deprecated `--kv_cache_type paged`
2. **Missing flags**: `--remove_input_padding` not specified
3. **Dimension mismatch**: Decoder's `--max_encoder_input_len=1500` should be `3000`

## ✅ Corrected Build Commands

### Encoder Build
```bash
trtllm-build \
    --checkpoint_dir ${checkpoint_dir}/encoder \
    --output_dir ${output_dir}/encoder \
    --paged_kv_cache enable \
    --remove_input_padding enable \
    --moe_plugin disable \
    --max_batch_size 8 \
    --gemm_plugin disable \
    --bert_attention_plugin float16 \
    --max_input_len 3000 \
    --max_seq_len 3000
```

### Decoder Build
```bash
trtllm-build \
    --checkpoint_dir ${checkpoint_dir}/decoder \
    --output_dir ${output_dir}/decoder \
    --paged_kv_cache enable \
    --remove_input_padding enable \
    --moe_plugin disable \
    --max_beam_width 2 \
    --max_batch_size 8 \
    --max_seq_len 114 \
    --max_input_len 14 \
    --max_encoder_input_len 3000 \
    --gemm_plugin float16 \
    --bert_attention_plugin float16 \
    --gpt_attention_plugin float16
```

### Key Changes
| Parameter | ❌ Original | ✅ Fixed |
|-----------|------------|---------|
| Encoder cache | `--kv_cache_type paged` | `--paged_kv_cache enable` |
| Encoder padding | *(missing)* | `--remove_input_padding enable` |
| Decoder cache | `--kv_cache_type paged` | `--paged_kv_cache enable` |
| Decoder padding | *(missing)* | `--remove_input_padding enable` |
| Decoder encoder len | `--max_encoder_input_len 1500` | `--max_encoder_input_len 3000` |

## 🧪 Testing

After building, test with:

```bash
# Download sample audio
wget -O sample_audio.wav \
    https://raw.githubusercontent.com/yuekaizhang/Triton-ASR-Client/main/datasets/mini_en/wav/1221-135766-0002.wav

# Run inference
python3 /path/to/TensorRT-LLM/examples/run.py \
    --engine_dir ${output_dir} \
    --multimodal_input_file sample_audio.wav \
    --max_output_len 96
```

## 📦 Complete Automated Solution

I've also created an automated build script that:
- Auto-detects TensorRT-LLM location
- Downloads Whisper models automatically
- Installs dependencies
- Builds with correct configuration

### Download the Script

```bash
# Get the automated build script
wget https://raw.githubusercontent.com/[your-repo]/whisper_encoder_guide/build_whisper_engines.sh
chmod +x build_whisper_engines.sh

# Run it
./build_whisper_engines.sh cpp
```

Or use this complete script:

```bash
#!/bin/bash
set -e

INFERENCE_PRECISION="float16"
WEIGHT_ONLY_PRECISION="int8"
MAX_BEAM_WIDTH=4
MAX_BATCH_SIZE=8
MODEL_NAME="large-v3"

checkpoint_dir="whisper_${MODEL_NAME}_weights_${WEIGHT_ONLY_PRECISION}"
output_dir="whisper_${MODEL_NAME}_${WEIGHT_ONLY_PRECISION}"

# Convert checkpoint (assuming you have the model downloaded)
python3 convert_checkpoint.py \
    --use_weight_only \
    --weight_only_precision ${WEIGHT_ONLY_PRECISION} \
    --output_dir ${checkpoint_dir} \
    --model_name ${MODEL_NAME}

# Build Encoder
trtllm-build \
    --checkpoint_dir ${checkpoint_dir}/encoder \
    --output_dir ${output_dir}/encoder \
    --paged_kv_cache enable \
    --remove_input_padding enable \
    --moe_plugin disable \
    --max_batch_size ${MAX_BATCH_SIZE} \
    --gemm_plugin disable \
    --bert_attention_plugin ${INFERENCE_PRECISION} \
    --max_input_len 3000 \
    --max_seq_len 3000

# Build Decoder
trtllm-build \
    --checkpoint_dir ${checkpoint_dir}/decoder \
    --output_dir ${output_dir}/decoder \
    --paged_kv_cache enable \
    --remove_input_padding enable \
    --moe_plugin disable \
    --max_beam_width ${MAX_BEAM_WIDTH} \
    --max_batch_size ${MAX_BATCH_SIZE} \
    --max_seq_len 114 \
    --max_input_len 14 \
    --max_encoder_input_len 3000 \
    --gemm_plugin ${INFERENCE_PRECISION} \
    --bert_attention_plugin ${INFERENCE_PRECISION} \
    --gpt_attention_plugin ${INFERENCE_PRECIPITATION}

echo "✓ Build complete!"
echo "Test with:"
echo "  python3 examples/run.py --engine_dir ${output_dir} --multimodal_input_file audio.wav --max_output_len 96"
```

## 🎯 Why This Fixes Your Errors

### Error 1: Dimension Mismatch (3000 != 1500)
**Fixed by**: Setting `--max_encoder_input_len 3000` in decoder to match encoder's `--max_input_len 3000`

### Error 2: Buffer Overflow (slice 1152000 exceeds 576000)
**Fixed by**: Adding `--paged_kv_cache enable` and `--remove_input_padding enable` for proper memory allocation

## 📊 Verified Configuration

This configuration is verified against the official TensorRT-LLM test suite:
- Source: [`tests/integration/defs/examples/test_whisper.py`](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/defs/examples/test_whisper.py)
- Uses C++ runtime with paged KV cache
- Supports all Whisper model sizes

## 💡 Tips

1. **Use C++ runtime** (with `paged_kv_cache enable`) for production - better performance
2. **For debugging**, you can use Python runtime:
   - Set `--paged_kv_cache disable` and `--remove_input_padding disable`
   - Add `--use_py_session` when running
3. **Memory issues?** Reduce `--max_batch_size` or use a smaller model (e.g., `medium`, `small`, `tiny`)

## 🔗 References

- [TensorRT-LLM Whisper README](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/models/core/whisper)
- [Encoder-Decoder Guide](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/models/core/enc_dec)

---

Let me know if you need any clarification or run into other issues!

**Comment by karljang** [comment]
@crazydemo , wow! thank you for your contribution!

@MahmoudAshraf97 , Could you please try out the new PR and let us know if it resolves the issue with the zero-value output tensor?

**Comment by karljang** [comment]
@ozancaglayan,

I believe the PR you found is #9842. However, please note that it hasn’t been merged due to latency concerns. More details are available in the [review comments](https://github.com/NVIDIA/TensorRT-LLM/pull/9842/changes/BASE..a024545a3c89263de93b3e1175660e18084b7d88).

Regarding the AI-generated answers, I also had concerns about them. Therefore, when I initiated the conversation, I explicitly mentioned this in [my comment](https://github.com/NVIDIA/TensorRT-LLM/issues/2338#issuecomment-3544136444).

> I’m sharing a potential solution below, generated by my AI tool for your reference.

Thanks!
